packages/inline-edit/src/InlineEdit.tsx

```diff
@@ -35,133 +29,152 @@ type InlineEditProps = ComponentProps<'div'> &
-        />
+      isEditing: isEditingProp,
+      onCancel,
+      onEdit,
+      cancelButtonLabel = 'cancel',
+      editButtonLabel = 'edit',
+      confirmButtonLabel = 'confirm',
+      className,
+      ...rest
+    },
+    ref
+  ) => {
+    const [isEditing, setEditing] = useState(isEditingProp ?? false);
+    const [isFocusWithin, setFocusWithin] = useState(false);
+    const inputRef = useRef<HTMLInputElement>(null);
+    const editRef = useRef<HTMLButtonElement>(null);
+    const controlled = isEditingProp !== undefined;
+
+    useUpdateEffect(() => {
+      if (controlled) {
+        setEditing(isEditingProp);
+      }
+    }, [isEditingProp]);
+
+    useUpdateEffect(() => {
+      if (isFocusWithin) {
+        isEditing
+          ? inputRef.current && focusSafely(inputRef.current)
+          : editRef.current && focusSafely(editRef.current);
+      }
+    }, [isEditing]);
+
+    const handleEdit = () => {
+      !controlled && setEditing(true);
+      onEdit?.();
+    };
+
+    const handleCancel = () => {
+      !controlled && setEditing(false);
+      onCancel?.();
+    };
+
+    const handleConfirm = () => {
+      onConfirm(inputRef.current?.value || '');
+      !controlled && setEditing(false);
+    };
+
+    const handleKeyDown: KeyboardEventHandler<HTMLInputElement> = (event) => {
+      if (event.key === 'Enter') {
+        event.preventDefault();
+        handleConfirm();
+      } else if (event.key === 'Escape') {
+        event.preventDefault();
+        handleCancel();
+      }
+    };
+
+    const { focusWithinProps } = useFocusWithin({
+      onBlurWithin: () => isEditing && handleCancel(),
+      onFocusWithinChange: (isFocusWithin) => setFocusWithin(isFocusWithin),
+    });
+
+    const { buttonProps } = useButton(
+      {
+        'aria-label': editButtonLabel,
+        elementType: 'span',
+        onPress: handleEdit,
+      },
+      editRef
+    );
+
+    const renderReadContent = hideEdit ? (
+      <span ref={editRef} {...buttonProps} className={readButton}>
+        {children}
+      </span>
+    ) : (
+      <>
+        {children}
         <IconButton
-          kind="default"
-          icon={<Icon name="close" />}
-          aria-label={cancelButtonLabel}
-          className={cancelButton}
-          onClick={handleCancel}
+          ref={editRef}
+          icon={<Icon name="edit" />}
+          aria-label={editButtonLabel}
+          size="small"
+          onClick={handleEdit}
         />
-      </ButtonGroup>
-    </div>
-  ) : (
-    <div className={cx(!hideEdit && container)} data-test-id={testId} {...focusWithinProps}>
-      {renderReadContent}
-    </div>
-  );
-};
+      </>
+    );
+
+    const input = cloneElement(
+      renderInput,
+      mergeProps(renderInput.props, {
+        ref: mergeRefs(inputRef, ref),
+        defaultValue,
+        onKeyDown: handleKeyDown,
+        'aria-label': ariaLabel,
+      })
+    );
+
+    return isEditing ? (
+      <div
+        {...rest}
```

Comment:
I assume these props will always map to valid DOM attributes, right?

Reply:
Yeah the ones coming from `ComponentProps<'div'>`.